### PR TITLE
#123 (best issue number): onLoggedIn API

### LINKED ISSUE #123: onLoggedIn API
it would be nice if there was some API for telling users to log in if not logged it
```typescript
declare function onLoggedIn(
    callback: () => void, // The thing to do if logged in
    promptContainerId: string, // The html element to stick a "You should log in" button in
    description?: string, // some description of what logging in will give you
);
```
If already logged in, just callback immediately
Else, on log in, call all registered callbacks and delete all buttons

Description useful if we implement #118

### PR DESCRIPTION
**Pull request by adirar111**
wip

resolve #123

### CODE REVIEW

**Comment by adirar111** [comment]
@braxtonhall 

should it work like this

bruh

File: extension/ts/main/extensions/util/onLoggedIn.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+function removeInjectedButton() {
+	const loginButton = document.getElementById(LOGIN_BUTTON_ID);
+	loginButton?.remove();
+}
+
+async function onLoggedIn(callback: () => void, container: HTMLElement) {
+	if (!(await isAuthorized())) {
+		const onClick = async () => {
+			await authorize(true);
```
Comment (braxtonhall) [inline extension/ts/main/extensions/util/onLoggedIn.ts:22]: I think we should add a `.then` and `.catch` to this call.

If it's successful then we remove the button and call the callback. Else we do nothing

File: extension/ts/main/extensions/util/onLoggedIn.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import {invokeWorker} from "../../../common/workers/invoker";
+import {WorkerKind} from "../../../common/workers/types";
+import {createIconButton} from "../../ui/button";
+import {isAuthorized} from "../author/util/isAuthorized";
+
+const LOGIN_BUTTON_ID = "vbl-bll-login";
```
Comment (braxtonhall) [inline extension/ts/main/extensions/util/onLoggedIn.ts:6]: I'm not sure I am a fan of this button having a specific ID.

The problem is then we can have at most one button per page. I think you could get around this but capturing the button in the scope of `removeInjectedButton`, could be defined in the scope of `onLoggedIn`

**Review by braxtonhall — COMMENTED** [review]
I think this is pretty air tight. Only a couple tiny changes and it's good to ship I'd say

File: extension/ts/main/extensions/util/onLoggedIn.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import {invokeWorker} from "../../../common/workers/invoker";
+import {WorkerKind} from "../../../common/workers/types";
+import {createIconButton} from "../../ui/button";
+import {isAuthorized} from "../author/util/isAuthorized";
+
+const authorize = (interactive: boolean) => invokeWorker(WorkerKind.Authorize, interactive).catch(() => "");
```
Comment (braxtonhall) [inline extension/ts/main/extensions/util/onLoggedIn.ts:6]: I think the catch on this line is preventing the other catch from ever executing

File: extension/ts/main/extensions/util/onLoggedIn.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+function injectButton(button: HTMLTableCellElement, container: HTMLElement) {
```
Comment (braxtonhall) [inline extension/ts/main/extensions/util/onLoggedIn.ts:8]: what happened to our beautiful arrow functions :(

File: extension/ts/main/adapters/author.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@ import Sheets, {GetSheetsDataResponse, ValueRange} from "./sheets";
 import {filterAuthorTags} from "../util/filterAuthorTags";
 import {makeCache} from "../util/cache";
 
-const SPREADSHEET_ID = "1EfwBhY56M8OwgVjFTWxxxdoIxK8osw2vfgsXnCyGGuA";
+const SPREADSHEET_ID = "18I5LabO21LfV97CkBRBW6SeK5hPggitvnK-2joUJ8jU";
```
Comment (braxtonhall) [inline extension/ts/main/adapters/author.ts:5]: UM can we not commit this?

**Review by braxtonhall — COMMENTED** [review]
ok last changes for real this time lol

File: extension/ts/main/adapters/author.ts (hunk 5 above)
Comment (adirar111) [reply extension/ts/main/adapters/author.ts:5]: AAAAAAAAAAAAAAA yeah mb

File: extension/ts/main/extensions/util/onLoggedIn.ts (hunk 4 above)
Comment (adirar111) [reply extension/ts/main/extensions/util/onLoggedIn.ts:8]: oop

File: extension/ts/main/extensions/util/onLoggedIn.ts (hunk 3 above)
Comment (adirar111) [reply extension/ts/main/extensions/util/onLoggedIn.ts:6]: oop

File: extension/ts/main/extensions/util/onLoggedIn.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	container.append(button);
+};
+
+const removeInjectedButton = (button: HTMLTableCellElement) => {
+	button.remove();
+};
+
+const onLoggedIn = async (callback: () => void, container: HTMLElement) => {
+	if (!(await isAuthorized())) {
+		const onClick = async () => {
```
Comment (braxtonhall) [inline extension/ts/main/extensions/util/onLoggedIn.ts:18]: ok nitpicking time. extra async.
also a maybe prettier version (very subjective)
```typescript
const onClick = () => 
	authorize(true)
		.then(() => removeInjectedButton(button))
		.then(callback)
		.catch(console.error);
```

**Review by braxtonhall — APPROVED** [review]
smh @ these commit names

**Comment by braxtonhall** [comment]
Question: what happens if you complete the sign in but you're signed in to an account that is not part of the Better LibraryThing GCP project?

**Pull request merged**